awx/main/tasks.py

```diff
@@ -372,16 +389,29 @@ def gather_analytics():
             if acquired is False:
                 logger.debug('Not gathering analytics, another task holds lock')
                 return
-            try:
-                tgz = analytics.gather()
-                if not tgz:
-                    return
-                logger.info('gathered analytics: {}'.format(tgz))
-                analytics.ship(tgz)
-                settings.AUTOMATION_ANALYTICS_LAST_GATHER = gather_time
-            finally:
-                if os.path.exists(tgz):
-                    os.remove(tgz)
+            subset = list(all_collectors().keys())
+            incremental_collectors = []
+            for collector in expensive_collectors():
+                if collector in subset:
+                    subset.remove(collector)
+                    incremental_collectors.append(collector)
+
+            # Cap gathering at 4 weeks of data if there has been no data gathering
+            since = last_time or (gather_time - timedelta(weeks=4))
+
+            if incremental_collectors:
+                start = since
+                until = None
+                while start < gather_time:
+                    until = start + timedelta(minutes=20)
+                    if (until > gather_time):
+                        until = gather_time
+                    if not _gather_and_ship(incremental_collectors, since=start, until=until):
+                        break
+                    start = until
+                    settings.AUTOMATION_ANALYTICS_LAST_GATHER = until
+            if subset:
+                _gather_and_ship(subset, since=since, until=gather_time)
```

Comment:
We should probably do `...last_gather = gather_time` here, so after everything is sent, to make sure we get all data in a case of failure somewhere in the middle.

(we'd need to store another timestamp, if we'd want to record incremental collectors and the rest of `subset`)

Reply:
So the way I intended to code this (I may have gotten it wrong) was to shift `settings.AUTOMATION_ANALYTICS_LAST_GATHER` to "the latest date we have job information for".

Hence we update it each time sending a time slice of the data succeeds, stop collecting that time sliced data entirely if one slice fails, and then send the non-automation-record data (counts, config,json, instance_info, etc) *without* updating the last gather date. 

This is a semantic change of what it represents, but as long as the data processor does the right thing, I think it's more accurate, and will ensure we (in theory) won't miss any job records.

Reply:
Ah ok, only the `incremental_collectors` are time constrained, so for the rest of the subset, the `AUTOMATION_ANALYTICS_LAST_GATHER` doesn't matter. So then this is correct. You can mark this as resolved. :-)